azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -167,6 +187,81 @@ def test_passes_event_to_previous_stage_which_throws_base_exception(
                 stage.handle_pipeline_event(event)
             assert unhandled_error_handler.call_count == 0
 
-            pass
-
     setattr(module, "Test{}UnknownEvents".format(cls.__name__), LocalTestObject)
+
+
+class ThreadLaunchedError(Exception):
+    pass
+
+
+def add_pipeline_thread_tests(
+    cls, module, methods_that_enter_pipeline_thread, methods_that_can_run_in_any_thread
+):
+    def does_method_assert_pipeline_thread(method_name):
+        if method_name.startswith("__"):
+            return False
+        elif method_name in methods_that_enter_pipeline_thread:
+            return False
+        elif method_name in methods_that_can_run_in_any_thread:
+            return False
+        else:
+            return True
+
+    methods_that_assert_pipeline_thread = [
+        x[0]
+        for x in inspect.getmembers(cls, inspect.isfunction)
+        if does_method_assert_pipeline_thread(x[0])
+    ]
+
+    @pytest.mark.describe("{} - Pipeline threading".format(cls.__name__))
+    class LocalTestObject(object):
+        @pytest.fixture
+        def stage(self):
+            return cls()
+
+        @pytest.mark.parametrize("method_name", methods_that_assert_pipeline_thread)
+        @pytest.mark.it("Has a method that must be in a pipeline thread")
+        def test_asserts_in_pipeline(self, stage, method_name, fake_non_pipeline_thread):
+            func = getattr(stage, method_name)
+            args = [None for i in (range(get_arg_count(func) - 1))]
+            with pytest.raises(AssertionError):
+                func(*args)
+
+        if methods_that_enter_pipeline_thread:
```

Comment:
This is wild, I didn't know you could put conditional statements inside a class. Very interesting.

Reply:
I thought you'd appreciate this.  This is another way that javascript has warped my fragile little mind -- instead of parametrize, the javascript test framework lets you just add the same test function multiple times inside a for loop, and that bends your brain until it doesn't.  

---
In reply to: [305485131](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r305485131) [](ancestors = 305485131)